bigquery_magics/config.py

```diff
@@ -26,7 +27,151 @@ def _get_default_credentials_with_project():
     return pydata_google_auth.default(scopes=_SCOPES, use_local_webserver=False)
 
 
-@dataclass
+class MissingReason(enum.Enum):
+    """Provides a way to disambiguate why an option is missing.
+
+    This is used instead of a None value to allow for custom validation and
+    docs generation.
+    """
+
+    # These missing reasons are because the engine doesn't support the feature,
+    # for example, use_rest_api on the bigframes engine. In this case, raise
+    # if the user has set it, as the magics would otherwise act in a way
+    # contrary to that in which the user explicitly requested.
+    NOT_SUPPORTED_BY_ENGINE_INFEASIBLE = enum.auto()
+
+    # This is like the above, but is theoretically possible. Include a call to
+    # action to reach out to the bigframes team if this is a feature they
+    # would like to use.
+    NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE = enum.auto()
+
+    # This missing reason is for options that are magics-only and apply to all
+    # engines. For example, the destination variable is a magics-only setting
+    # and doesn't affect how queries are handled by the engine.
+    ENGINE_UNIVERSAL = enum.auto()
+
+
+@dataclasses.dataclass(frozen=True)
+class MagicsSetting:
+    """Encapsulates information about settings and how to set them.
+
+    This is used to generate documentation, merge settings across the various
+    ways they are duplicated, and to validate settings provided by a user.
+    """
+
+    description: str
+    cell_arg: Union[MissingReason, str]
+    magics_context: Union[MissingReason, str]
+    bigframes_option: Union[MissingReason, str]
+
+
+magics_settings = [
+  MagicsSetting(
+      description="Limits the number of rows in the returned DataFrame.",
+      cell_arg="--max_results",
+      magics_context=MissingReason.NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE,
+      bigframes_option="read_gbq_query.max_results"
+  ),
+
+  MagicsSetting(
+      description="Max bytes billed to use for executing a query.",
+      cell_arg="--maximum_bytes_billed",
+      magics_context="default_query_job_config",
+      bigframes_option=MissingReason.NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE
+  ),
+
+  MagicsSetting(
+      description="Set it to a query to estimate costs.",
+      cell_arg="--dry_run",
+      magics_context=MissingReason.NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE,
+      bigframes_option="read_gbq_query"
+  ),
+
+  MagicsSetting(
+      description="Set it to use instead of Standard SQL.",
+      cell_arg="--use_legacy_sql",
+      magics_context=MissingReason.NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE,
```

Comment:
We will almost never have `NOT_SUPPORTED_BY_ENGINE_BUT_POSSIBLE` for `magics_context` because that was the first engine we created (pandas engine). It's the bigframes one that is missing features like this.

If there is something missing, `SUPPORTED_BY_ENGINE_BUT_NO_OPTION` is likely the reason.